### DISCUSSION TITLE
TypeError: stat: path should be string, bytes, os.PathLike or integer, not NoneType tfc_end2end_example

### DISCUSSION THREAD

**Discussion by sakshigoel361** [Xilinx/finn#442] [Q&A] [answered]
Hi,
I see similar error as discussed in the thread https://github.com/Xilinx/finn/discussions/392
I am running the example bnn-pynq/tfc_end2end_example inside the docker using 'bash ./run-docker.sh notebook'. I am using the latest version of dev branch. While using the function DeployToZYNQ, i get the error about static path. If I take care of this using os.stat(target_dir), then the error changes to that directory '/home/xilinx/finn_tfc_end2end_example' does not exist.
I am able to ssh to PYNQ board with bash ./run_docker.sh and cd to the directory /home/xilinx/finn_tfc_end2end_example , so it does exist. I have created this directory using mkdir before calling DeployToZYNQ
Error1
![Q3](https://user-images.githubusercontent.com/38778510/142641388-c4b6c808-3662-432b-81e6-663623977af3.png)
Error2
![Screenshot from 2021-11-19 15-09-09](https://user-images.githubusercontent.com/38778510/142641422-d6ae24bb-5ff9-45eb-b4b0-63d95e3381ad.png)

**Comment by rpitonak** [comment, marked as answer]
Hello, 

I think the problem may be that on FINN dev there is no PYNQ driver generated by default during the `Zynq build` and these notebooks were not adapted to this change (I can do a PR once the answer is accepted and it resolves this issue). Try to add this before the step you are running.

```python
from finn.core.modelwrapper import ModelWrapper
from finn.transformation.fpgadataflow.make_pynq_driver import MakePYNQDriver

model = ModelWrapper(build_dir + "/tfc_w1_a1_post_synthesis.onnx")
model = model.transform(MakePYNQDriver(platform="zynq-iodma"))
model.save(build_dir + "/tfc_w1_a1_pynq-driver.onnx")
```

and then execute

```python
from finn.transformation.fpgadataflow.make_deployment import DeployToPYNQ

model = ModelWrapper(build_dir + "/tfc_w1_a1_pynq-driver.onnx")
model = model.transform(DeployToPYNQ(ip, port, username, password, target_dir))
model.save(build_dir + "/tfc_w1_a1_pynq_deploy.onnx")
```

Hope it helps.

**Reply by sakshigoel361** [reply]
@rpitonak Thank you for your help. With the addition you suggested, the error indeed goes away. I am using ZCU104 and on running the further steps of the tutorial I realized that the board does not have PYNQ image (There is no /pynq folder and I get the error ModuleNotFoundError: No module name 'pynq' when executing driver.py script on the board). 
I think after adding a PYNQ image maybe additional lines of code are not needed. The board has ZCU104 image from previous VitisAI experiments and in order to add PYNQ image maybe I need to flash the SD card again with PYNQ image. Is this image sufficient for FINN  dev?

**Reply by rpitonak** [reply]
Yes, the preferred target platform for FINN is a board with PYNQ image.

I am not using ZCU104, but it should be supported by FINN out-of-the-box since it is defined in `pynq_part_map`:
https://github.com/Xilinx/finn-base/blob/dev/src/finn/util/basic.py

For the PYNQ setup of your board see these links (just for reference, you probably already found them)
SD card image: http://www.pynq.io/board.html
Getting started guide: https://pynq.readthedocs.io/en/v2.6.1/getting_started/zcu104_setup.html#zcu104-board-setup

**Reply by sakshigoel361** [reply]
@rpitonak The example works now till the end and I can try to modify it for my project. Thanks for your help :)

**Reply by superpi15** [reply]
My PYNQ-Z2 works well on end to end example now. 
Thank you again for the solution.
